src/Common/FST.cpp

```diff
@@ -0,0 +1,445 @@
+
+            previous_written_bytes = minimized_state->serialize(write_buffer);
+            state_count++;
+            previous_state_index += previous_written_bytes;
+        }
+    }
+}
+
+void FSTBuilder::add(const std::string& current_word, Output current_output)
+{
+    /// We assume word size is no greater than MAX_TERM_LENGTH
+    auto current_word_len = current_word.size();
+
+    if (current_word_len > MAX_TERM_LENGTH)
+        throw Exception(DB::ErrorCodes::LOGICAL_ERROR, "Too long term ({}) passed to FST builder.", current_word_len);
+
+    size_t prefix_length_plus1 = getCommonPrefix(current_word, previous_word) + 1;
+
+    minimizePreviousWordSuffix(prefix_length_plus1);
+
+    /// Initialize the tail state
+    for (size_t  i = prefix_length_plus1; i <= current_word.size(); ++i)
+    {
+        temp_states[i]->clear();
+        temp_states[i - 1]->addArc(current_word[i-1], 0, temp_states[i]);
+    }
+
+    /// We assume the current word is different with previous word
+    temp_states[current_word_len]->flag_values.is_final = true;
+    /// Adjust outputs on the arcs
+    for (size_t j = 1; j <= prefix_length_plus1 - 1; ++j)
+    {
+        auto * arc_ptr = temp_states[j - 1]->getArc(current_word[j-1]);
+        assert(arc_ptr != nullptr);
+
+        auto common_prefix = std::min(arc_ptr->output, current_output);
+        auto word_suffix = arc_ptr->output - common_prefix;
+        arc_ptr->output = common_prefix;
+
+        /// For each arc, adjust its output
+        if (word_suffix != 0)
+        {
+            for (auto& label_arc : temp_states[j]->arcs)
+            {
+                auto& arc = label_arc.second;
+                arc.output += word_suffix;
+            }
+        }
+        /// Reduce current_output
+        current_output -= common_prefix;
+    }
+
+    /// Set last temp state's output
+    auto * arc = temp_states[prefix_length_plus1 - 1]->getArc(current_word[prefix_length_plus1-1]);
+    assert(arc != nullptr);
+    arc->output = current_output;
+
+    previous_word = current_word;
+}
+
+UInt64 FSTBuilder::build()
+{
+    minimizePreviousWordSuffix(0);
+
+    /// Save initial state index
+
+    previous_state_index -= previous_written_bytes;
+    UInt8 length = getLengthOfVarUInt(previous_state_index);
+    writeVarUInt(previous_state_index, write_buffer);
+    write_buffer.write(length);
+
+    return previous_state_index + previous_written_bytes + length + 1;
+}
+
+FST::FST(std::vector<UInt8>&& data_) : data(data_)
+{
+}
+
+void FST::clear()
+{
+    data.clear();
+}
+
+std::pair<bool, UInt64> FST::getOutput(const String& term)
+{
+    std::pair<bool, UInt64> result_output{ false, 0 };
+    /// Read index of initial state
+    ReadBufferFromMemory read_buffer(data.data(), data.size());
+    read_buffer.seek(data.size()-1, SEEK_SET);
+
+    UInt8 length{0};
+    read_buffer.read(reinterpret_cast<char&>(length));
+
+    read_buffer.seek(data.size() - 1 - length, SEEK_SET);
```

Comment:
It is kinds of difficult to validate value of length since it can be any value in theory. But we will keep it in mind to find more general way for validation.